server/src/internalClusterTest/java/org/elasticsearch/snapshots/GetSnapshotsIT.java

```diff
@@ -208,6 +209,67 @@ public void testPaginationRequiresVerboseListing() throws Exception {
         );
     }
 
+    public void testExcludePatterns() throws Exception {
+        final String repoName1 = "test-repo-1";
+        final String repoName2 = "test-repo-2";
+        final String otherRepo = "other-repo";
+        createRepository(repoName1, "fs");
+        createRepository(repoName2, "fs");
+        createRepository(otherRepo, "fs");
+
+        final List<String> namesRepo1 = createNSnapshots(repoName1, randomIntBetween(1, 5));
+        final List<String> namesRepo2 = createNSnapshots(repoName2, randomIntBetween(1, 5));
+        final List<String> namesOtherRepo = createNSnapshots(otherRepo, randomIntBetween(1, 5));
+
+        final Collection<String> allSnapshotNames = new HashSet<>(namesRepo1);
+        allSnapshotNames.addAll(namesRepo2);
+        final Collection<String> allSnapshotNamesWithoutOther = Set.copyOf(allSnapshotNames);
+        allSnapshotNames.addAll(namesOtherRepo);
+
+        final SortOrder order = SortOrder.DESC;
+        final List<SnapshotInfo> allSorted = allSnapshotsSorted(allSnapshotNames, "*", GetSnapshotsRequest.SortBy.REPOSITORY, order);
+        final List<SnapshotInfo> allSortedWithoutOther = allSnapshotsSorted(
+            allSnapshotNamesWithoutOther,
+            "*,-" + otherRepo,
+            GetSnapshotsRequest.SortBy.REPOSITORY,
+            order
+        );
+        assertThat(allSorted.subList(0, allSnapshotNamesWithoutOther.size()), is(allSortedWithoutOther));
+
+        final List<SnapshotInfo> allInOther = allSnapshotsSorted(
+            namesOtherRepo,
+            "*,-test-repo-*",
+            GetSnapshotsRequest.SortBy.REPOSITORY,
+            order
+        );
+        assertThat(allSorted.subList(allSnapshotNamesWithoutOther.size(), allSorted.size()), is(allInOther));
+
+        final String otherPrefixSnapshot1 = "other-prefix-snapshot-1";
+        createFullSnapshot(otherRepo, otherPrefixSnapshot1);
+        final String otherPrefixSnapshot2 = "other-prefix-snapshot-2";
+        createFullSnapshot(otherRepo, otherPrefixSnapshot2);
+
+        final String patternOtherRepo = randomBoolean() ? otherRepo : "*,-test-repo-*";
+        final List<SnapshotInfo> allInOtherWithoutOtherPrefix = allSnapshotsSorted(
+            namesOtherRepo,
+            patternOtherRepo,
+            GetSnapshotsRequest.SortBy.REPOSITORY,
+            order,
+            "-other*"
+        );
+        assertThat(allInOtherWithoutOtherPrefix, is(allInOther));
+
+        final List<SnapshotInfo> allInOtherWithoutOtherExplicit = allSnapshotsSorted(
+            namesOtherRepo,
+            patternOtherRepo,
+            GetSnapshotsRequest.SortBy.REPOSITORY,
+            order,
+            "-" + otherPrefixSnapshot1,
+            "-" + otherPrefixSnapshot2
+        );
+        assertThat(allInOtherWithoutOtherExplicit, is(allInOther));
```

Comment:
Should we test the result when a snapshot is both included and excluded?

Reply:
++ added explicit check for this case